src/wiktextract/extractor/de/gloss.py

```diff
@@ -3,58 +3,148 @@
-                wxr.wtp.debug(
-                    f"Skipped a sub-list in gloss list: {sub_list_node}",
-                    sortid="extractor/de/glosses/extract_glosses/27",
-                )
-                # XXX: We should extract the subglosses as subsenses.
-                pass
+
+            # Extract sub-glosses for later processing
+            sub_glosses_list_nodes = list(
+                find_and_remove_child(list_item_node, NodeKind.LIST)
+            )
+
+            raw_gloss = clean_node(wxr, {}, list_item_node.children)
+            gloss_data["raw_glosses"] = [raw_gloss]
+
+            extract_categories_from_gloss_node(wxr, gloss_data, list_item_node)
 
             gloss_text = clean_node(wxr, gloss_data, list_item_node.children)
 
-            match = re.match(r"\[(\d+[a-z]?)\]", gloss_text)
-            if match:
-                sense_number = match.group(1)
-                gloss_text = gloss_text[match.end() :].strip()
-            else:
-                sense_number = None
+            senseid, gloss_text = match_senseid(gloss_text)
 
-            if not sense_number:
+            if senseid:
+                senseid = (
+                    senseid
+                    if senseid[0].isnumeric()
+                    else parent_senseid + senseid
+                )
+                gloss_data["senseid"] = senseid
+            else:
                 wxr.wtp.debug(
-                    f"Failed to extract sense number from gloss: {gloss_text}",
+                    f"Failed to extract sense number from gloss node: {list_item_node}",
                     sortid="extractor/de/glosses/extract_glosses/28",
                 )
 
-            gloss_data["glosses"] = [gloss_text]
+            gloss_text = extract_categories_from_gloss_text(
+                gloss_data, gloss_text
+            )
+
+            if gloss_text or not sub_glosses_list_nodes:
+                gloss_data["glosses"] = [gloss_text]
+                page_data[-1]["senses"].append(gloss_data)
 
-            page_data[-1]["senses"].append(gloss_data)
+            for sub_list_node in sub_glosses_list_nodes:
+                process_gloss_list_item(
+                    wxr,
+                    page_data,
+                    sub_list_node,
+                    senseid,
+                    gloss_data if not gloss_text else None,
+                )
 
         else:
             wxr.wtp.debug(
                 f"Unexpected list item in glosses: {list_item_node}",
                 sortid="extractor/de/glosses/extract_glosses/29",
             )
             continue
+
+
+def handle_sense_modifier(wxr, list_item_node):
+    wxr.wtp.debug(
+        f"Skipped a sense modifier in gloss list: {list_item_node}",
+        sortid="extractor/de/glosses/extract_glosses/19",
+    )
+    # XXX: We should extract the modifier. However, it seems to affect
+    # multiple glosses. Needs investigation.
+    pass
+
+
+def extract_categories_from_gloss_node(
```

Comment:
But generally the German Wiktionary doesn't use links to the category namespace to modify single senses. So the `clean_node()` function doesn't help here.

It mostly uses the [K](https://de.wiktionary.org/wiki/Vorlage:K) template (literally: context template) which sometimes creates an entry in a category page, sometimes just outputs information about the context of usage (which we might rather store in the "`tags`" field).

Anyway I think we need some custom logic here to capture the contextual data present in the German Wiktionary.


_As an aside: I find the `clean_node()` function a bit obscure and hard to work with. I am basically using it only to create clean text from a node. To me it would seem cleaner if all other side-effects would be moved to a separate function. But perhaps I am missing some deeper reason here._

Reply:
Isn't the "K" template also expands to a category link? `clean_node` will find all the category links in the expanded text and save them to the passed object. It should work for the German Wiktionary. But if it doesn't work for some cases, you're free to use a new implantation.

I also only use it to convert node to text and save category data, I guess that's the main usage of this function. It needs better documentation...